module_utils/network_lsr/nm/active_connection.py

```diff
@@ -0,0 +1,127 @@
+from ansible.module_utils.network_lsr.nm.client import get_client  # noqa:E501
+from ansible.module_utils.network_lsr.nm.error import LsrNetworkNmError  # noqa:E501
+
+# pylint: enable=import-error, no-name-in-module
+
+
+NM_AC_STATE_CHANGED_SIGNAL = "state-changed"
+
+
+def deactivate_active_connection(nm_ac, timeout, check_mode):
+    if not nm_ac or nm_ac.props.state == NM.ActiveConnectionState.DEACTIVATED:
+        logging.info("Connection is not active, no need to deactivate")
+        return False
+    if not check_mode:
+        main_loop = get_mainloop(timeout)
+        logging.debug(
+            "Deactiving {id} with timeout {timeout}".format(
+                id=nm_ac.get_id(), timeout=timeout
+            )
+        )
+        user_data = main_loop
+        handler_id = nm_ac.connect(
+            NM_AC_STATE_CHANGED_SIGNAL, _nm_ac_state_change_callback, user_data
+        )
+        logging.debug(
+            "Registered {signal} on NM.ActiveConnection {id}".format(
+                signal=NM_AC_STATE_CHANGED_SIGNAL, id=nm_ac.get_id()
+            )
+        )
+        if nm_ac.props.state != NM.ActiveConnectionState.DEACTIVATING:
+            nm_client = get_client()
+            user_data = (main_loop, nm_ac, handler_id)
+            nm_client.deactivate_connection_async(
+                nm_ac,
+                main_loop.cancellable,
+                _nm_ac_deactivate_call_back,
+                user_data,
+            )
+            logging.debug("Deactivating NM.ActiveConnection {0}".format(nm_ac.get_id()))
+        main_loop.run()
+    return True
+
+
+def _nm_ac_state_change_callback(nm_ac, state, reason, main_loop):
+    if main_loop.is_cancelled:
+        return
+    logging.debug(
+        "Got NM.ActiveConnection state change: {id}: {state} {reason}".format(
+            id=nm_ac.get_id(), state=state, reason=reason
+        )
+    )
+    if nm_ac.props.state == NM.ActiveConnectionState.DEACTIVATED:
+        logging.debug("NM.ActiveConnection {0} is deactivated".format(nm_ac.get_id()))
+        main_loop.quit()
+
+
+def _nm_ac_deactivate_call_back(nm_client, result, user_data):
+    main_loop, nm_ac, handler_id = user_data
+    logging.debug("NM.ActiveConnection deactivating callback")
+    if main_loop.is_cancelled:
+        if nm_ac:
+            nm_ac.handler_disconnect(handler_id)
+        return
+
+    try:
+        success = nm_client.deactivate_connection_finish(result)
+    except GLib.Error as e:
+        if e.matches(NM.ManagerError.quark(), NM.ManagerError.CONNECTIONNOTACTIVE):
+            success = True
+            logging.info(
+                "Connection is not active on {0}, no need to deactivate".format(
+                    nm_ac.get_id()
+                )
+            )
+            if nm_ac:
+                nm_ac.handler_disconnect(handler_id)
+            main_loop.quit()
+        else:
+            _deactivate_fail(
```

Comment:
there is a race here. The client calls `deactivate_connection_async()` (after checking that the active connection is still <= ACTIVATED), but still, at that very moment (on the server side) the ActiveConnection could go down and disappear entirely. Then it wouldn't merely fail with `NM.ManagerError.CONNECTIONNOTACTIVE`, but with some entirely different message like that the D-Bus path does not refer to an existing object.

(of course, it's a race and maybe hard to reproduce)

Reply:
If you are referring the race of `nm_ac.get_id()`, I have fixed it by saving the id in the user data.

For other race issue on this `some entirely different message`, never seen in nmstate 1000+ test runs. I don't know the detail and why it could failed, so I will live with it.